File: low-level-design/design-patterns/golang/behavioural/state-pattern/state/restarting_state.go
package state

import (
	"errors"
	"fmt"
)

func NewRestartingState() *RestartingState {
	return &RestartingState{}
}

type RestartingState struct {
}

func (s *RestartingState) Start(vm *VM) error {
	fmt.Printf("%s vm is already restarting \n",vm.name)
	return errors.New("vm is already restarting")
}

func (s *RestartingState) Stop(vm *VM) error {
	fmt.Printf("%s stopping the vm \n",vm.name)
	vm.SetState(NewStoppedState())
	return nil
}
func (s *RestartingState) Restart(vm *VM) error {
	fmt.Printf("%s vm is already restarting \n",vm.name)
	return errors.New("vm is already restarting")
}
func (s *RestartingState) Suspend(vm *VM) error {
	fmt.Printf("%s suspending the vm \n",vm.name)
	vm.SetState(NewSuspendedState())
	return nil
}
func (s *RestartingState) GetName() string {
	return "Restarting"
}

File: low-level-design/design-patterns/python/behavioural/state-pattern/restarting_state.py
from state import VMState





class RestartingError(Exception):
    pass 


class RestartingState(VMState):
    def start(self, vm):
        from state_factory import StateFactory
        print(f"Starting VM {vm.name}")
        vm.set_state(StateFactory.get_starting_state())
    
    def stop(self, vm):
        from state_factory import StateFactory
        print(f"Stopping VM {vm.name}")
        vm.set_state(StateFactory.get_stopped_state())
    
    def restart(self, vm):
        print(f"VM {vm.name} is already in restarting state, no action performed")
        return RestartingError("VM is already in restarting state, no action performed")
    
    def suspend(self, vm):
        from state_factory import StateFactory
        print(f"Suspending VM {vm.name}")
        vm.set_state(StateFactory.get_suspended_state())
    
    def get_name(self):
        return "Restarting"